# Machine Learning Project

## predicting house prices using regression

# Problem Statement

## Problems Faced during buying a house:
### 1.Buying house is stressful thing
### 2.Buyers are not generally aware of the factors that influence the house price
### 3.Many problems are faced during buying a house
### 4.Hence real estate agents trusted with communication between seller and buyer as well as laying down a legal contract. This just creates another person in between and increases the cost


# Problem Objective

### 1.Our project is a machine learning model based on certain specifications of your home it will try to guess the most accurate price.
### 2.Information such as location, number of bedrooms etc
### 3. To predict the sale price of each houses


# Importing Libraries

In [178]:
import pandas as pd
import numpy as np

In [179]:
import os 
os.getcwd()

'C:\\Users\\Phaneendra\\Desktop'

## Reading data

In [180]:
traindata=pd.read_csv(r'C:\Users\Phaneendra\Downloads\wk3_kc_house_train_data.csv\wk3_kc_house_train_data.csv')

In [181]:
validationdata=pd.read_csv(r'C:\Users\Phaneendra\Downloads\wk3_kc_house_valid_data.csv\wk3_kc_house_valid_data.csv')

In [182]:
traindata.columns #columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [183]:
traindata.head()

id             date      price  bedrooms  bathrooms  sqft_living  \
0  2487200875  20141209T000000   604000.0         4       3.00         1960   
1  7237550310  20140512T000000  1225000.0         4       4.50         5420   
2  9212900260  20140527T000000   468000.0         2       1.00         1160   
3   114101516  20140528T000000   310000.0         3       1.00         1430   
4  6054650070  20141007T000000   400000.0         3       1.75         1370   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5000     1.0           0     0  ...      7        1050            910   
1    101930     1.0           0     0  ...     11        3890           1530   
2      6000     1.0           0     0  ...      7         860            300   
3     19901     1.5           0     0  ...      7        1430              0   
4      9680     1.0           0     0  ...      7        1370              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1965             0    98136  47.5208 -122.393           1360   
1      2001             0    98053  47.6561 -122.005           4760   
2      1942             0    98115  47.6900 -122.292           1330   
3      1927             0    98028  47.7558 -122.229           1780   
4      1977             0    98074  47.6127 -122.045           1370   

   sqft_lot15  
0        5000  
1      101930  
2        6000  
3       12697  
4       10208  

[5 rows x 21 columns]

In [184]:
traindata.dtypes #datatype of data present in each column

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

In [185]:
traindata.shape #shape

(9761, 21)

In [186]:
validationdata.shape #shape

(9635, 21)

## checking for null values

In [187]:
traindata.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

## modelling with selected features(highly coorelated columns)

In [188]:
train=traindata[corr_cols]
test=validationdata[corr_cols]
y=traindata[['price']]
z=validationdata[['price']]

# modelling

In [189]:
def results(train,test):
    from sklearn.linear_model import LinearRegression
    model1=LinearRegression()
    model1.fit(train,y)
    pred1=model1.predict(test)
    from sklearn.metrics import mean_squared_error           #importing mean square error metric
    import numpy as np
    mse=mean_squared_error(z,pred1)
    rmse=np.sqrt(mse)
    print(f'rmse: {rmse}')
    from sklearn.metrics import r2_score                    # importing r2 score metric
    score=r2_score(z,pred1)
    print(f'score: {score}')


In [190]:
results(train,test)                    #rmse and r2 score for model    

rmse: 230245.05643559116
score: 0.5821056743107639


###### when modelling is done with the selecte features based on correalation factor we acheived rmse of 230245 and r2 score of 0.58

## modeling with each and every feature

In [191]:
train=traindata.drop(['id','price','date'],axis=1)              #removing the id,target variable 
test=validationdata.drop(['id','price','date'],axis=1)

In [192]:
traindata.head()  

id             date      price  bedrooms  bathrooms  sqft_living  \
0  2487200875  20141209T000000   604000.0         4       3.00         1960   
1  7237550310  20140512T000000  1225000.0         4       4.50         5420   
2  9212900260  20140527T000000   468000.0         2       1.00         1160   
3   114101516  20140528T000000   310000.0         3       1.00         1430   
4  6054650070  20141007T000000   400000.0         3       1.75         1370   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5000     1.0           0     0  ...      7        1050            910   
1    101930     1.0           0     0  ...     11        3890           1530   
2      6000     1.0           0     0  ...      7         860            300   
3     19901     1.5           0     0  ...      7        1430              0   
4      9680     1.0           0     0  ...      7        1370              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1965             0    98136  47.5208 -122.393           1360   
1      2001             0    98053  47.6561 -122.005           4760   
2      1942             0    98115  47.6900 -122.292           1330   
3      1927             0    98028  47.7558 -122.229           1780   
4      1977             0    98074  47.6127 -122.045           1370   

   sqft_lot15  
0        5000  
1      101930  
2        6000  
3       12697  
4       10208  

[5 rows x 21 columns]

In [193]:
results(train,test)                       #rmse and r2 score for model

rmse: 194970.288941156
score: 0.7003441806096995


##### when modelling is done with  selecting each and every feature and without data preprocessing  we acheived rmse of 19470 and r2 score of 0.70

## data preprocessing

In [194]:
traindata1=traindata.copy()                     #creating a copy
validationdata1=validationdata.copy()                

In [195]:
traindata1['age']=2015-traindata1['yr_built']       #from yr_built column we are calculating age
traindata1=traindata1.drop('yr_built',axis=1)

In [196]:
validationdata1['age']=2015-validationdata1['yr_built'] #from yr_built column we are calculating age
validationdata1=validationdata1.drop('yr_built',axis=1)

In [197]:
traindata1.yr_renovated[traindata1['yr_renovated']>0]=1 #changing the yr_renovated column

C:\Users\Phaneendra\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [198]:
validationdata1.yr_renovated[validationdata1['yr_renovated']>0]=1       #changing the yr_renovated column

C:\Users\Phaneendra\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


## modelling with including each and every column and preprocessed data

In [199]:
train=traindata1.drop(['id','date','price'],axis=1)
test=validationdata1.drop(['id','date','price'],axis=1)

In [200]:
results(train,test)                        #rmse and r2score of model

rmse: 194975.6770662321
score: 0.7003276180313351


##### when modelling is done with the selecting each and every feature and changing some of the features we acheived rmse of 195239 and r2 score of 0.69 which is an improvement to the above

# using log transformation

In [205]:
train['sqft_living']=np.log(train['sqft_living'])
train['sqft_above']=np.log(train['sqft_above'])
train['sqft_lot']=np.log(train['sqft_lot'])
train['sqft_lot15']=np.log(train['sqft_lot15'])
train['sqft_living15']=np.log(train['sqft_living15'])
train=train.drop(['lat','long'],axis=1)


In [206]:
y=np.log(y)

In [207]:
test['sqft_living']=np.log(test['sqft_living'])
test['sqft_above']=np.log(test['sqft_above'])
test['sqft_lot']=np.log(test['sqft_lot'])
test['sqft_lot15']=np.log(test['sqft_lot15'])
test['sqft_living15']=np.log(test['sqft_living15'])
test=test.drop(['lat','long'],axis=1)

In [208]:
from sklearn.linear_model import LinearRegression
model1=LinearRegression()
model1.fit(train,y)
pred1=model1.predict(test)

In [209]:
pred1=np.exp(pred1)

In [210]:
from sklearn.metrics import mean_squared_error           #importing mean square error metric
import numpy as np
mse=mean_squared_error(z,pred1)
rmse=np.sqrt(mse)
print(f'rmse: {rmse}')
from sklearn.metrics import r2_score                    # importing r2 score metric
score=r2_score(z,pred1)
print(f'score: {score}')

rmse: 120816.10682286728
score: 0.8849370459942972


##### when modelling is done with the selecting each and every feature and converting zipcode using dummies method  and applying log transformation we acheived rmse of 124427 and r2 score of 0.87 which is an improvement to the above

# converting zipcode using dummies

In [201]:
cols=pd.get_dummies(traindata1['zipcode'])           #converting categorical to numerical
traindata1=pd.concat([traindata1,cols],axis=1)
traindata1=traindata1.drop('zipcode',axis=1)

In [202]:
cols1=pd.get_dummies(validationdata1['zipcode'])            #converting categorical to numerical
validationdata1=pd.concat([validationdata1,cols1],axis=1)
validationdata1=validationdata1.drop('zipcode',axis=1)

In [203]:
train=traindata1.drop(['id','date','price'],axis=1)
test=validationdata1.drop(['id','date','price'],axis=1)

##### when modelling is done with the selecting each and every feature and converting zipcode using dummies method we acheived rmse of 158832 and r2 score of 0.80 which is not an improvement to the above

In [204]:
results(train,test)                           #rmse and r2score of the model

rmse: 157284.42368566606
score: 0.8049898230660651


In [229]:
import matplotlib.pyplot as plt                         #plotting rmse values and model
model=['model1','model2','model3','model4']
rmse=[230245,194975,157284,120816]
plt.xlabel('models')
plt.ylabel('rmse')
plt.plot(model,rmse)
plt.scatter(x=model,y=rmse)
plt.show()

##### By observing above graph we cay say that model 6 has low rmse and it is the best model so we will apply model 6 on the test data for the better results

##   model 1 -> rmse 230245   r2 score ->0.58
##   model 2 -> rmse 194970   r2 score ->0.70
##   model 3 -> rmse 195239   r2 score ->0.69
##   model 4 -> rmse 158830   r2 score ->0.80
##   model 6 -> rmse 123905   r2 score ->0.87

# Reading test data

In [211]:
testdata=pd.read_csv(r'C:\Users\Phaneendra\Downloads\wk3_kc_house_test_data.csv\wk3_kc_house_test_data.csv')

In [212]:
testdata.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  3793500160  20150312T000000  323000.0         3       2.50         1890   
1  1175000570  20150312T000000  530000.0         5       2.00         1810   
2    16000397  20141205T000000  189000.0         2       1.00         1200   
3   461000390  20140624T000000  687500.0         4       1.75         2330   
4  7895500070  20150213T000000  240000.0         4       1.00         1220   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      6560     2.0           0     0  ...      7        1890              0   
1      4850     1.5           0     0  ...      7        1810              0   
2      9850     1.0           0     0  ...      7        1200              0   
3      5000     1.5           0     0  ...      7        1510            820   
4      8075     1.0           0     0  ...      7         890            330   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      2003             0    98038  47.3684 -122.031           2390   
1      1900             0    98107  47.6700 -122.394           1360   
2      1921             0    98002  47.3089 -122.210           1060   
3      1929             0    98117  47.6823 -122.368           1460   
4      1969             0    98001  47.3341 -122.282           1290   

   sqft_lot15  
0        7570  
1        4850  
2        5095  
3        5000  
4        7800  

[5 rows x 21 columns]

In [213]:
testdata.shape

(2217, 21)

# data preptocessing

In [214]:
testdata['age']=2015-testdata['yr_built']            #creating age column from yr_built column
testdata=testdata.drop('yr_built',axis=1)

In [215]:
testdata.yr_renovated[testdata['yr_renovated']>0]=1          #converting yr_renovated column

C:\Users\Phaneendra\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [216]:
testdata['sqft_living']=np.log(testdata['sqft_living'])
testdata['sqft_above']=np.log(testdata['sqft_above'])
testdata['sqft_lot']=np.log(testdata['sqft_lot'])
testdata['sqft_lot15']=np.log(testdata['sqft_lot15'])
testdata['sqft_living15']=np.log(testdata['sqft_living15'])
testdata=testdata.drop(['lat','long'],axis=1)

In [217]:
cols2=pd.get_dummies(testdata['zipcode'])                               #converting categorical to numerical
testdata=pd.concat([testdata,cols2],axis=1)
testdata=testdata.drop('zipcode',axis=1)

In [218]:
p=testdata[['price']]
test=testdata.drop(['id','date','price'],axis=1)

In [219]:
from sklearn.linear_model import LinearRegression
model=LinearRegression()                                                 #fitting the model
model.fit(train,y)
pred=model.predict(test)                                                 #predicting

In [220]:
pred=np.exp(pred)

In [221]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
mse=mean_squared_error(p,pred)
score=r2_score(p,pred)
print(f'rmse:{np.sqrt(mse)}')                            #rmse value for model

rmse:126847.45385717908


##### when we applied our best model on test data we acheived rmse of 126847